hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/statemachine/background/BlockDeletingService.java

```diff
@@ -312,9 +338,102 @@ public BackgroundTaskResult call() throws Exception {
         }
         crr.addAll(succeedBlocks);
         return crr;
-      } finally {
-        container.writeUnlock();
+      } catch (IOException exception) {
+        LOG.warn(
+            "Deletion operation was not successful for container: " + container
+                .getContainerData().getContainerID());
+      }
+      return crr;
+    }
+
+    public ContainerBackgroundTaskResult deleteViaSchema2(
+        ReferenceCountedDB meta, Container container, File dataDir,
+        long startTime) {
+      ContainerBackgroundTaskResult crr = new ContainerBackgroundTaskResult();
+      try {
+        Table<String, BlockData> blockDataTable =
+            meta.getStore().getBlockDataTable();
+        DatanodeStore ds = meta.getStore();
+        DatanodeStoreSchemaTwoImpl dnStoreTwoImpl =
+            (DatanodeStoreSchemaTwoImpl) ds;
+        Table<Long, DeletedBlocksTransaction>
+            delTxTable = dnStoreTwoImpl.getTxnTable();
+        List<DeletedBlocksTransaction> delBlocks = new ArrayList<>();
+        int totalBlocks = 0;
+        try (TableIterator<Long,
+            ? extends Table.KeyValue<Long, DeletedBlocksTransaction>> iter =
+            dnStoreTwoImpl.getTxnTable().iterator()) {
+          while (iter.hasNext() && (totalBlocks < blockLimitPerTask)) {
```

Comment:
Thanks for pointing this out @errose28. It can violate the blockLimitPerTask here if totalBlocks + delTx.getLocalIDList().size() > blockLimitPerTask but this would handle a case when a transaction contains blocks > blockLimitPerTask. If we would put a check to blockLimitPerTask inside while loop then the transaction which contains blocks > blockLimitPerTask will be completely ignored and will never be deleted. We also have some tests in testBlockDeletingService.java where the container contains blocks more than blockLimitPerTask.

Reply:
Ah yes, I see that now. It seems SCM's `DeletedBlockLogImpl` takes a similar approach, where if the first transaction has more blocks than the block deleting limit, or the last transaction added puts the number of blocks over the limit, it allows it and it will be forwarded to the datanode. Although this isn't a super strict interpretation of the block deleting limit, it seems consistent with the rest of the code so I think it is ok.